nuvlaedge/peripherals/modbus/__init__.py

```diff
@@ -76,35 +98,58 @@ def manage_modbus_peripherals(ip_address):
     #    modbus.{port}.{interface}.{identifier}
 
     # Ask the NB agent for all modbus peripherals matching this pattern
+    # probably the kubernetes stuff should go here?
+    # why not?
+
     logger.info(f'Starting modbus scan on {ip_address}')
     xml_file = scan_open_ports(ip_address)
 
     parser = NmapOutputXMLParser(xml_file)
     parser.parse()
 
     all_modbus_devices = parser.get_modbus_details()
+    logger.debug(f"All modbus devices:\n {all_modbus_devices}")
+
     discovered_devices: dict = {}
-    if ip_address not in all_modbus_devices:
-        logger.warning(f'No Modbus Info found for host : {ip_address}')
-        return discovered_devices
-
-    for per in all_modbus_devices[ip_address]:
-        port = per.get("port", "nullport")
-        interface = per.get("interface", "nullinterface")
-        identifiers = per.get("identifiers").copy()
-        del per["identifiers"]
-        for ident in identifiers:
-            _id = ident['key']
-            per['classes'] = ident['classes']
-            per['vendor'] = ident['vendor']
-            per['name'] = ident['name']
-            identifier = "modbus.{}.{}.{}".format(port, interface, _id)
-            # Redefine the identifier
-            per['identifier'] = identifier
-            discovered_devices[identifier] = per.copy()
+
+    for returned_ip_address in all_modbus_devices.keys():
+        for per in all_modbus_devices[returned_ip_address]:
+            port = per.get("port", "nullport")
+            interface = per.get("interface", "nullinterface")
+            identifiers = per.get("identifiers").copy()
+            del per["identifiers"]
+            for ident in identifiers:
+                _id = ident['key']
+                per['classes'] = ident['classes']
+                per['vendor'] = ident['vendor']
+                per['name'] = ident['name']
+                per['testhost'] = returned_ip_address
+                identifier = "host {}:{} interface {} id {}".format(returned_ip_address, port, interface, _id)
+                # Redefine the identifier
+                per['identifier'] = identifier
+                discovered_devices[identifier] = per.copy()
+
+    logger.debug(f"discovered devices:\n {discovered_devices}")
 
     return discovered_devices
 
+def determine_ip_addresses(list_of_ip_addresses):
+    """Determine the list of IP addresses to be scanned."""
+
+    command = "ip -4 -br a s"
+    logger.debug(f'IP address command:\n{command}')
+
+    known_ip_addresses = os.system(command)
+
+    logger.debug(f"Known IP addresses:\n {known_ip_addresses}")
```

Comment:
Do we get the list of IPs of the system only to show a debug message ?
If yes: can this whole code block be removed ?
If no: then there is an bug because the variable `known_ip_addresses` is not used except in the log message.

If this code block should be kept, it should be refactored to use a function from `subprocess` module to execute system command instead of `os.system` and possible exceptions should be handled.

Reply:
No, it's not just a debug...  The list of IP addresses is passed on to the line 170.
Indeed I can change the way the IP is called.

Reply:
My point wasn't to remove the whole function, only the code bloc corresponding to the lines 140 to 145.
I wasn't talking about `list_of_ip_addresses` but about `known_ip_addresses`.

Can you please re-evalute my first comment/questions with these additional explanations.

Reply:
Oh yes... you are right that known_ip_addresses goes nowhere. I updated the function to test for routable IP addresses on the host.